br/pkg/lightning/backend/external/engine.go

```diff
@@ -58,3 +63,200 @@ func (e *Engine) createMergeIter(ctx context.Context, start kv.Key) (*MergeKVIte
 	}
 	return iter, nil
 }
+
+// MemoryIngestData is the in-memory implementation of IngestData.
+type MemoryIngestData struct {
+	keyAdapter         local.KeyAdapter
+	duplicateDetection bool
+	duplicateDB        *pebble.DB
+	dupDetectOpt       local.DupDetectOpt
+
+	keys   [][]byte
+	values [][]byte
+	ts     uint64
+}
+
+var _ local.IngestData = (*MemoryIngestData)(nil)
+
+func (m *MemoryIngestData) firstAndLastKeyIndex(lowerBound, upperBound []byte) (int, int) {
+	firstKeyIdx := 0
+	if len(lowerBound) > 0 {
+		lowerBound = m.keyAdapter.Encode(nil, lowerBound, local.MinRowID)
+		firstKeyIdx = sort.Search(len(m.keys), func(i int) bool {
+			return bytes.Compare(lowerBound, m.keys[i]) <= 0
+		})
+		if firstKeyIdx == len(m.keys) {
+			return -1, -1
+		}
+	}
+
+	lastKeyIdx := len(m.keys) - 1
+	if len(upperBound) > 0 {
+		upperBound = m.keyAdapter.Encode(nil, upperBound, local.MinRowID)
+		i := sort.Search(len(m.keys), func(i int) bool {
+			reverseIdx := len(m.keys) - 1 - i
+			return bytes.Compare(upperBound, m.keys[reverseIdx]) > 0
+		})
+		if i == len(m.keys) {
+			// should not happen
+			return -1, -1
+		}
+		lastKeyIdx = len(m.keys) - 1 - i
+	}
+	return firstKeyIdx, lastKeyIdx
+}
+
+// GetFirstAndLastKey implements IngestData.GetFirstAndLastKey.
+func (m *MemoryIngestData) GetFirstAndLastKey(lowerBound, upperBound []byte) ([]byte, []byte, error) {
+	firstKeyIdx, lastKeyIdx := m.firstAndLastKeyIndex(lowerBound, upperBound)
+	if firstKeyIdx < 0 || firstKeyIdx > lastKeyIdx {
```

Comment:
Maybe return an error when `firstKeyIdx > lastKeyIdx` (the keys in `MemoryIngestData` are not sorted)?

Reply:
in firstAndLastKeyIndex we use sort.Search which relies on the keys are sorted. If the keys are not sorted, I'm afraid "firstKeyIdx > lastKeyIdx" will not always fail. Do you think we can add a quick check about m.keys[0] is less than m.keys[len(m.keys)-1]?

Reply:
> I'm afraid "firstKeyIdx > lastKeyIdx" will not always fail

OK. I think current implementation is fine enough.